File: examples/ios-sdk-examples/ios-sdk-examples-objc/Views/POI/POIViewController.m
//
//  POIViewController.m
//  ios-sdk-examples-objc
//
//  Created by iMacbook on 2/1/21.
//

#import "POIViewController.h"
#import <Map4dMap/Map4dMap.h>
#import <UIKit/UIKit.h>

@interface POIViewController ()

@property (weak, nonatomic) IBOutlet MFMapView *mapView;
@property (weak, nonatomic) IBOutlet UIButton *removePOIButton;
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;

@end

@interface POIViewController(MFMapViewDelegate) <MFMapViewDelegate>
@end

@implementation POIViewController
{
  NSMutableArray<MFPOI*> *pois;
  int i;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  [[self navigationItem] setTitle: @"POI"];
  [_mapView setPOIsEnabled: NO];
  [_mapView setDelegate: self];
  CLLocationCoordinate2D location = _mapView.camera.target;
  pois = [NSMutableArray array];
  i = 0;
  [pois addObject:[self drawPOILocation:location title:@"MAP4D" type:@"cafe"]] ;
  
}

- (IBAction)didTapAdd:(id)sender {
  [_removePOIButton setHidden: NO];
  [_titleLabel setHidden: YES];
  
  double lat = (random() % 500) / 50000.0 - 0.005;
  double lng = (random() % 500) / 50000.0 - 0.005;
  CLLocationCoordinate2D location = _mapView.camera.target;
  location.latitude = location.latitude + lat;
  location.longitude = location.longitude + lng;
  i += 1;
  
  MFPOI * poi = [self drawPOILocation:location title:[NSString stringWithFormat:@"MAP4D - %d", i] type:@"cafe"];
  [pois addObject:poi] ;
}

- (IBAction)didTapRemove:(id)sender {
  [_titleLabel setHidden: YES];
  NSUInteger count = [pois count];
  if (count < 1) {
    return;
  }
  [_removePOIButton setHidden: count < 2];
  [[pois lastObject] setMap: nil];
  [pois removeLastObject];
  
}

- (MFPOI*) drawPOILocation: (CLLocationCoordinate2D) location title: (NSString*) title type: (NSString*) type {
  MFPOI* poi = [[MFPOI alloc] init];
  [poi setPosition: location];
  [poi setTitle: title];
  [poi setType: type];
  [poi setZIndex: 2.0];
  [poi setTitleColor: UIColor.redColor];
  [poi setMap: _mapView];
  return poi;
}
@end

@implementation POIViewController(MFMapViewDelegate)
- (void)mapView:(MFMapView *)mapView didTapPOI:(MFPOI *)poi {
  [_titleLabel setHidden: NO];
  [_titleLabel setText: [NSString stringWithFormat: @"title: %@", poi.title]];
}

- (void)mapView:(MFMapView *)mapView didTapPOIWithPlaceID:(NSString *)placeID name:(NSString *)name location:(CLLocationCoordinate2D)location {
  [_titleLabel setHidden: NO];
  [_titleLabel setText: [NSString stringWithFormat: @"title: %@", name]];
}
@end

File: examples/ios-sdk-examples/ios-sdk-examples-swift/Views/Home/poi/POIViewController.swift
//
//  POIViewController.swift
//  ios-sdk-examples
//
//  Created by iMacbook on 1/27/21.
//  Copyright © 2021 iot. All rights reserved.
//

import UIKit
import Map4dMap

class POIViewController: UIViewController {
  
  @IBOutlet weak var mapView: MFMapView!
  @IBOutlet weak var removePOIButton: UIButton!
  @IBOutlet weak var titleLabel: UILabel!
  var pois: [MFPOI] = []
  var i = 0
  
  override func viewDidLoad() {
    super.viewDidLoad()
    mapView.setPOIsEnabled(false)
    self.navigationItem.title = "POI"
    mapView.delegate = self
    guard let location = mapView.camera?.target else {return}
    pois.append(draw(location: location, title: "Map4D", type: "point"))
    
  }
  
  @IBAction func didTapAddPOI(_ sender: Any) {
    let detal = Double.random(in: 0.0...0.0099999) - 0.0099999 / 2
    let detal1 = Double.random(in: 0.0...0.0099999) - 0.0099999 / 2
    let types = ["cafe", "atm", "point"]
    let type = types[Int.random(in: 0...2)]
    i += 1
    removePOIButton.isHidden = false
    guard let location = mapView.camera?.target else {return}
    
    pois.append(draw(location: CLLocationCoordinate2D(
          latitude: location.latitude + detal,
          longitude: location.longitude + detal1), title: "POI\(i)", type: type))
  }
  
  @IBAction func didTapRemovePOI(_ sender: Any) {
    removePOIButton.isHidden = !(i != 0)
    guard i >= 0 else {return}
    pois[i].map = nil
    pois.remove(at: i)
    i -= 1
  }
  
  func draw(location: CLLocationCoordinate2D, title: String?, type: String?) -> MFPOI{
    let poi = MFPOI()
    poi.position = location
    print(location)
    poi.title = title
    poi.zIndex = 2
    poi.type = type
    poi.titleColor = .red
    poi.map = mapView
    return poi
  }
  
}
extension POIViewController: MFMapViewDelegate {
  func mapView(_ mapView: MFMapView!, didTapPOIWithPlaceID placeID: String!, name: String!, location: CLLocationCoordinate2D) {
    titleLabel.isHidden = false
    titleLabel?.text = "ID: \(placeID ?? "") NAME: \(name ?? "")"
  }
  func mapView(_ mapView: MFMapView!, didTap poi: MFPOI!) {
    titleLabel.isHidden = false
    titleLabel?.text = "ID: \(poi.id) NAME: \(poi.title ?? "")"
  }
}
